src/osx/cocoa/dnd.mm

```diff
@@ -29,6 +31,217 @@
+    DoSetData( kUTTypeFileURL, data);
+}
+
+wxOSXDataSourceItem::~wxOSXDataSourceItem()
+{
+}
+
+bool wxOSXDataSource::IsSupported(const wxDataFormat &dataFormat)
+{
+    wxCFMutableArrayRef<CFStringRef> typesarray;
+    dataFormat.AddSupportedTypes(typesarray);
+    return HasData(typesarray);
+}
+
+bool wxOSXDataSource::IsSupported(const wxDataObject &dataobj)
+{
+    wxCFMutableArrayRef<CFStringRef> typesarray;
+    dataobj.AddSupportedTypes(typesarray);
+    return HasData(typesarray);
+}
+
+class WXDLLIMPEXP_CORE wxOSXPasteboardSinkItem : public wxOSXDataSinkItem
+{
+public:
+    wxOSXPasteboardSinkItem(NSPasteboardItem* item): m_item(item)
+    {
+    }
+
+    ~wxOSXPasteboardSinkItem()
+    {
+
+    }
+
+    virtual void SetData(const wxDataFormat& format, const void *buf, size_t datasize)
+    {
+        SetData( format.GetFormatId(), buf, datasize);
+    }
+
+    virtual void SetData(wxDataFormat::NativeFormat format, const void *buf, size_t datasize)
+    {
+        wxCFRef<CFDataRef> data(CFDataCreate( kCFAllocatorDefault, (UInt8*)buf, datasize ));
+        DoSetData(format, data);
+    }
+
+    virtual void DoSetData(wxDataFormat::NativeFormat format, CFDataRef data)
+    {
+        [m_item setData:(NSData*) data forType:(NSString*) format];
+    }
+
+    NSPasteboardItem* GetNative() { return m_item; };
+private:
+    NSPasteboardItem* m_item;
+};
+
+class WXDLLIMPEXP_CORE wxOSXPasteboardSourceItem : public wxOSXDataSourceItem
+{
+public:
+    wxOSXPasteboardSourceItem(NSPasteboardItem* item, NSPasteboard* board): m_item(item), m_pasteboard(board)
+    {
+    }
+
+    virtual wxDataFormat::NativeFormat AvailableType(CFArrayRef types) const
+    {
+        return (wxDataFormat::NativeFormat)[m_item availableTypeFromArray:(NSArray*)types];
+    }
+
+    virtual bool GetData( const wxDataFormat& dataFormat, wxMemoryBuffer& target)
+    {
+        return GetData(dataFormat.GetFormatId(), target);
+    }
+
+
+    virtual bool GetData( wxDataFormat::NativeFormat type, wxMemoryBuffer& target)
+    {
+        bool success = false;
+
+        target.Clear();
+
+        CFDataRef flavorData = DoGetData( type );
+
+        if ( flavorData )
+        {
+            CFIndex flavorDataSize = CFDataGetLength(flavorData);
+            size_t allocatedSize = flavorDataSize+4;
+            void * buf = target.GetWriteBuf(allocatedSize);
+            if ( buf )
+            {
+                memset(buf, 0, allocatedSize);
+                memcpy(buf, CFDataGetBytePtr(flavorData), flavorDataSize);
+                target.UngetWriteBuf(flavorDataSize);
+                success = true;
+            }
+        }
+        return success;
+    }
+
+    virtual CFDataRef DoGetData(wxDataFormat::NativeFormat type) const
+    {
+        // before a file promise can be resolved, we must pass a paste location
+        if ( UTTypeConformsTo((CFStringRef)type, kPasteboardTypeFileURLPromise ) )
```

Comment:
I don't understand why this is done or what it does; the comment leaves me as confused as I was before reading it.

Reply:
ok, I'll try to find a good link of docu, unfortunately the old docs which were quite good are gone for good ... basically a file promise is a type of data that promises to deliver a file, only if the data is really needed will it be created. This is used eg when dragging an attachment from an outlook mail. You as the client have then to provide a directory path, where outlook should place the file into, outlook then copies the file there and sets the data of this promise to be the new path.